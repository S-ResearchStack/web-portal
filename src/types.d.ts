declare module '@transcend-io/conflux';
declare module 'native-file-system-adapter';

interface ShowSaveFilePickerOptions {
  suggestedName: string;
}

interface Window {
  showSaveFilePicker: (options?: ShowSaveFilePickerOptions) => Promise<FileSystemFileHandle>;
}

declare module '*.svg' {
  import * as React from 'react';

  export const ReactComponent: React.FunctionComponent<
    React.SVGProps<SVGSVGElement> & { title?: string }
  >;

  export default ReactComponent;
}

interface CSSStyleDeclaration {
  webkitOverflowScrolling: string;
}

declare module '*.svg?url' {
  const src: string;
  export default src;
}

declare namespace NodeJS {
  interface ProcessEnv {
    NODE_ENV: 'production' | 'development' | 'test' | 'none';
    API_URL: string;
    SUPERSET_URL: string;
    // TODO: need to move to aws parameter store
    SUPERSET_ID: string;
    SUPERSET_PASSWORD: string;
    GOOGLE_CLIENT_ID: string;
    AUTHENTICATION_MODE: string;
    GOOGLE_OAUTH_URL: string;
    PUBLIC_PATH: string | undefined;
    MOCK_API: string;
    VERSION: string;
  }
}

declare module 'smoothscroll-polyfill' {
  class Polyfill {
    static polyfill(): void;
  }
  export default Polyfill;
}

declare module 'css-spring';

declare module "*.jpg";
declare module "*.png";
