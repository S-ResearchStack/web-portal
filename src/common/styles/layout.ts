import styled from "styled-components";

export const RowContainer = styled.div<{ flex?: number, width?: string, height?: string }>`
  display: flex;
  flex-direction: row;
  flex: ${p => p.flex ?? 1};
  width: ${p => p.width};
  height: ${p => p.height};
  flex-shrink: 0;
`

export const ColumnContainer = styled.div<{ width?: string, height?: string }>`
  display: flex;
  flex-direction: column;
  width: ${p => p.width};
  height: ${p => p.height};
`

export const PaddedColumnContainer = styled(ColumnContainer)`
  padding-left: 8px;
  padding-right: 8px;
`

export const SizedColumnContainer = styled(PaddedColumnContainer)`
  width: 100%;
`

export const SizedColumnScrollContainer = styled(SizedColumnContainer)`
  width: 100%;
  overflow: hidden;
  overflow-y: auto;
`

export const SizedBox = styled.div<{width?: number, height?: number}>`
  width: ${p => p.width ?? 0};
  height: ${p => p.height ?? 0};
  min-width: ${p => p.width ?? 0};
  min-height: ${p => p.height ?? 0};
  max-width: ${p => p.width ?? 0};
  max-height: ${p => p.height ?? 0};
`
