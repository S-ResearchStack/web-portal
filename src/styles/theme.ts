import { css, DefaultTheme } from 'styled-components';
import { px } from './utils';

const colors = {
  // gray gradations
  black: '#000000',
  black60: 'rgba(0, 0, 0, 0.6)',
  black40: 'rgba(0, 0, 0, 0.4)',
  black08: 'rgba(0, 0, 0, 0.08)',
  black15: 'rgba(0, 0, 0, 0.15)',

  // blue gradations
  blue: "#b0c4f1",
  blue40: 'rgba(176, 196, 241, 0.4)',
  blue15: 'rgba(176, 196, 241, 0.15)',
  blue8: 'rgba(176, 196, 241, 0.08)',

  // light mode
  surface: '#FFFFFF',
  background: '#F7F8FA',
  onPrimary: '#FFFFFF',
  backgroundOnPrimary: '#FFFFFF',
  backgroundSurface: '#FFFFFF',
  backgroundOnSurface: '#2D2D2D',
  primary: '#4475E3',
  primaryShadow: '#4475E361',
  primaryLight: '#ECF1FC',
  primaryHovered: '#2F65E0',
  primary05: '#F6F8FE',
  primary10: '#ECF1FC',
  primary20: '#E3EAFB',
  primary30: '#C7D6F7',
  primaryDisabled: '#B3C6F1',
  primaryLightPressed: '#E3EAFB',
  primaryLightFocused: '#C7D6F7',
  textPrimary: '#474747',
  primaryWhite: '#FFFFFF',
  disabled: '#DBDBDB',
  disabled20: 'rgba(219, 219, 219, 0.2)',
  onDisabled: '#A1A1A1',

  textPrimaryBlue: '#4475E3',
  textPrimaryBlue10: 'rgba(68, 117, 227, 0.2)',
  textPrimaryDark: '#474747',
  textSecondaryGray: '#808080',
  textDisabled: '#838383',
  onSurface: '#2D2D2D',
  secondarySkyblue: '#00B0D7', // replace
  backgoundLight: '#F0F0F0',
  primaryBlueHovered: '#2F65E0',
  primaryBluePressed: '#2057D5',

  statusSuccess: '#2DC008',
  statusSuccess10: '#DCF2EA',
  statusSuccessText: '#00825D',

  statusError: '#FF3333',
  statusError10: '#FFEBEB',
  statusErrorText: '#D14343',

  statusWarning: '#F5BF00',
  statusWarning10: '#FFF7D0',
  statusWarningText: '#A36400',

  secondaryViolet: '#944ED7',
  secondaryVioletHovered: '#6A3799',
  secondaryViolet10: '#F4EDFB',

  secondaryTangerine: '#F39C18',
  secondaryTangerineHovered: '#CC8314',
  secondaryTangerine10: '#FEF5E8',

  secondarySkyBlue: '#00B0D7',
  secondarySkyBlueHovered: '#0092B3',
  secondarySkyBlue10: '#E5F7FB',

  secondaryGreen: '#1AD598',
  secondaryGreenHovered: '#16B37F',
  secondaryGreen10: '#E8FBF5',

  secondaryRed: '#FA5F4F',
  secondaryRedHovered: '#CC4E40',
  secondaryRed10: '#FFEFED',

  transparent: 'rgba(0, 0, 0, 0)'
};

export type SpecColorType = keyof typeof colors;

const weightToCssValue = {
  regular: 400,
  upperregular: 450,
  medium: 500,
  semibold: 600,
  bold: 700,
};

const fontSpec = (
  size: number,
  lineHeight: number | string,
  weight: keyof typeof weightToCssValue,
  fontFamily: string
) => css`
  font-family: ${fontFamily}, sans-serif;
  font-size: ${px(size)};
  line-height: ${typeof lineHeight === 'string' ? lineHeight : px(lineHeight)};
  font-weight: ${weightToCssValue[weight]};
  font-style: normal;
`;

const typography = {
  // Inter typeface
  labelRegular: fontSpec(10, 15, 'regular', 'Inter'),
  labelSemibold: fontSpec(10, 15, 'semibold', 'Inter'),
  smallLabelRegular: fontSpec(8, 10, 'regular', 'Inter'),

  bodyXSmallRegular: fontSpec(12, 18, 'regular', 'Inter'),
  bodyXSmallSemibold: fontSpec(12, 18, 'semibold', 'Inter'),
  bodySmallRegular: fontSpec(14, 21, 'regular', 'Inter'),
  bodySmallSemibold: fontSpec(14, 21, 'semibold', 'Inter'),
  bodyMediumRegular: fontSpec(16, 24, 'regular', 'Inter'),
  bodyMediumSemibold: fontSpec(16, 24, 'semibold', 'Inter'),
  bodyLargeRegular: fontSpec(18, 26, 'regular', 'Inter'),
  bodyLargeSemibold: fontSpec(18, 26, 'semibold', 'Inter'),

  headingXSmall: fontSpec(16, 21, 'semibold', 'Inter'),
  headingSmall: fontSpec(18, 23, 'semibold', 'Inter'),
  headingXMedium: fontSpec(18, 27, 'semibold', 'Inter'),
  headingXMediumRegular: fontSpec(18, 27, 'regular', 'Inter'),
  headingMedium: fontSpec(24, 32.2, 'semibold', 'Inter'),
  headingLargeSemibold: fontSpec(32, 42, 'semibold', 'Inter'),
  heading2LargeSemibold: fontSpec(40, 52, 'semibold', 'Inter'),
  headingXLargeSemibold: fontSpec(48, 58, 'semibold', 'Inter'),
  headingXXLargeSemibold: fontSpec(60, 72, 'semibold', 'Inter'),

  portalDisplayLarge: fontSpec(72, 86, 'semibold', 'Inter'),

  sdkBodySmallRegular: fontSpec(14, '130%', 'regular', 'Inter'),
  sdkBodyLargeRegular: fontSpec(18, 24, 'regular', 'Inter'),
  sdkBodyLargeSemibold: fontSpec(18, 24, 'semibold', 'Inter'),
  sdkHeadingMedium: fontSpec(20, 26, 'semibold', 'Inter'),
  sdkBodyXXLRegular: fontSpec(32, 42, 'regular', 'Inter'),
  sdkBodyXXLSemibold: fontSpec(32, '130%', 'semibold', 'Inter'),
  sdkBodyMediumRegular: fontSpec(16, '130%', 'regular', 'Inter'),
  sdkBodyMediumSemibold: fontSpec(16, '130%', 'semibold', 'Inter'),
  sdkDisplayMedium: fontSpec(20, 26, 'semibold', 'Inter'),
  sdkHeadingLarge: fontSpec(40, '130%', 'semibold', 'Inter'),

  // Fira Code typeface
  query14: fontSpec(14, '130%', 'upperregular', 'Fira Code'),
  query12: fontSpec(12, '150%', 'upperregular', 'Fira Code'),
};

const boxShadow = {
  card: `0 0 ${px(2)} ${px(0)} rgba(0,0,0,0.15)`,
  avatar: `0 ${px(8)} ${px(10)} ${px(2)} 0 rgba(0,0,0,0.12)`,
  studyAvatar: `0 ${px(15)} ${px(20)} rgba(180,180,180,0.3)`,
  previewScreen: `0 ${px(2)} ${px(4)} rgba(71,71,71,0.25)`,
  skipLogicHeader: `box-shadow: ${px(3)} ${px(4)} ${px(15)} rgba(0, 0, 0, 0.05)`,
};

const animation = {
  defaultTiming: 'cubic-bezier(0.45, 0.05, 0, 1)',
  rippleTiming: 'cubic-bezier(0.4, 0, 0.2, 1)',
};

declare module 'styled-components' {
  export interface DefaultTheme {
    colors: typeof colors;
    typography: typeof typography;
    boxShadow: typeof boxShadow;
    animation: typeof animation;
  }
}

const theme: DefaultTheme = {
  colors,
  typography,
  boxShadow,
  animation,
};

export default theme;
