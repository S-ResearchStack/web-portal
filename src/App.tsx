import React from 'react';
import { Provider } from 'react-redux';
import { Route, Switch } from 'react-router-dom';
import { ConnectedRouter } from 'connected-react-router';
import { DndProvider } from 'react-dnd';
import { TouchBackend } from 'react-dnd-touch-backend';

import { ThemeProvider } from 'styled-components';

import { store } from 'src/modules/store/store';
import { history, Path } from 'src/modules/navigation/store';
import UserProtectedRoute from 'src/modules/navigation/private-route/UserProtectedRoute';
import MainLayout from 'src/modules/main-layout/MainLayout';
import SignIn from 'src/modules/auth/signin/SignIn';
import SignUp from 'src/modules/auth/signup/SignUp';
import { TooltipProvider, TooltipsList } from 'src/common/components/Tooltip';
import { theme, GlobalStyles } from 'src/styles';

import 'src/modules/auth/authProvider';
import { ModalProvider } from 'src/common/components/Modal';
import buildProvidersTree from './common/components/ProviderTree/ProviderTree';
import RegistrationScreen from 'src/modules/auth/registration/RegistrationScreen';
import TokenProtectedRoute from 'src/modules/navigation/private-route/TokenProtectedRoute';
import CreateStudyScreen from 'src/modules/studies/CreateStudyScreen';


const ProvidersTree = buildProvidersTree([
  [ThemeProvider, { theme: theme }],
  [TooltipProvider, {}],
  [ModalProvider, {}],
  [DndProvider, { backend: TouchBackend, options: { enableMouseEvents: true } }],
  [Provider, { store }]
]);

const App = () => (
  <ProvidersTree>
    <GlobalStyles />
    <ConnectedRouter history={history}>
      <Switch>
        <Route path={Path.SignIn} component={SignIn} />
        <Route path={Path.AccountCreate} component={SignUp} />
        <TokenProtectedRoute path={Path.Registration}>
          <RegistrationScreen />
        </TokenProtectedRoute>
        <UserProtectedRoute path={Path.CreateStudy}>
            <CreateStudyScreen />
        </UserProtectedRoute>
        <UserProtectedRoute>
          <MainLayout />
        </UserProtectedRoute>
      </Switch>
    </ConnectedRouter>
    <TooltipsList />
  </ProvidersTree>
);

export default App;
