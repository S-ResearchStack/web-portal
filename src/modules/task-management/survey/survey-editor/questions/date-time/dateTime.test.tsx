import React from 'react';
import '@testing-library/jest-dom';
import 'jest-styled-components';
import { act, render, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { ThemeProvider } from 'styled-components/';
import { TouchBackend } from 'react-dnd-touch-backend';
import { DndProvider } from 'react-dnd';
import { theme } from 'src/styles';
import handler from './index';

/* eslint-disable @typescript-eslint/no-explicit-any */

describe('date-time', () => {
  const expectId = expect.any(String);

  const emptyQuestion = {
    type: 'date-time',
    id: expectId,
    title: '',
    description: '',
    answers: [],
    options: {
      optional: false,
      isRange: false,
    },
    config: { isDate: true, isTime: true },
  };

  const notEmptyQuestion = {
    id: 'id',
    type: 'date-time',
    title: 'Title',
    description: 'Description',
    options: {
      optional: false,
      isRange: false,
    },
    config: { isDate: true, isTime: false },
  };

  const taskItemFromApi = {
    id: 'id',
    type: 'DATETIME',
    tag: 'DATETIME',
    title: 'Question Title',
    explanation: 'Question Description',
    required: true,
    itemProperties: {
      isDate: true,
      isTime: false,
      isRange: false,
    },
  };

  it('type should be date-time', async () => {
    expect(handler.type).toEqual('date-time');
  });

  it('[NEGATIVE] should return empty question', async () => {
    expect(handler.createEmpty()).toEqual(emptyQuestion);
  });

  it('isEmpty should return true', async () => {
    expect(handler.isEmpty(emptyQuestion as any)).toBeTrue();
  });

  it('[NEGATIVE] isEmpty should return true if question is undefined', async () => {
    expect(handler.isEmpty(undefined as any)).toBeTrue();
  });

  it('[NEGATIVE] isEmpty should return true if question answers is undefined', async () => {
    expect(handler.isEmpty({ ...emptyQuestion, answers: undefined } as any)).toBeTrue();
  });

  it('isEmpty should return false', async () => {
    expect(handler.isEmpty(notEmptyQuestion as any)).toBeFalse();
  });

  it('convertFromOtherType should return converted question', async () => {
    const qi = {
      type: 'single',
      id: 'Question1',
      title: 'Title',
      description: 'Description',
      options: {
        optional: false,
        includeOther: false,
      },
      answers: [{ answer1: 'answer1' }, { answer2: 'answer2' }],
    };
    expect(handler.convertFromOtherType(qi as any)).toEqual({
      ...qi,
      type: 'date-time',
      options: {
        optional: false,
        isRange: false,
      },
      answers: [],
      config: {
        isDate: true,
        isTime: true,
      },
    });
  });

  it('[NEGATIVE] convertFromOtherType should return empty question', async () => {
    expect(handler.convertFromOtherType({} as any)).toEqual({
      type: 'date-time',
      id: undefined,
      title: undefined,
      description: undefined,
      answers: [],
      options: {
        optional: false,
        isRange: false,
      },
      config: {
        isDate: true,
        isTime: true,
      },
    });
  });

  it('fromApi should return converted question', async () => {
    expect(handler.fromApi(taskItemFromApi as any)).toEqual({
      type: 'date-time',
      id: expectId,
      title: 'Question Title',
      description: 'Question Description',
      answers: [],
      options: {
        optional: false,
        isRange: false,
      },
      config: {
        isDate: true,
        isTime: false,
      },
    });
  });

  it('[NEGATIVE] fromApi should return undefined if question type is not an dropdown', async () => {
    expect(
      handler.fromApi({
        ...taskItemFromApi,
        type: '',
      } as any)
    ).toBeUndefined();
  });

  it('toApi should return converted question', async () => {
    expect(handler.toApi(notEmptyQuestion as any)).toEqual({
      id: expectId,
      type: 'DATETIME',
      tag: 'DATETIME',
      title: 'Title',
      explanation: 'Description',
      required: true,
      itemProperties: {
        isDate: true,
        isTime: false,
        isRange: false,
      },
    });
  });

  it('[NEGATIVE] toApi should not crash with empty data', async () => {
    expect(handler.toApi({} as any)).toEqual({
      id: undefined,
      type: 'DATETIME',
      tag: 'DATETIME',
      title: undefined,
      explanation: undefined,
      required: false,
      itemProperties: {
        isDate: undefined,
        isTime: undefined,
        isRange: undefined,
      },
    });
  });

  it('renderEditorContent should render editor content', async () => {
    const onChange = jest.fn();
    const confirmOptionRemoval = jest.fn();

    const { baseElement, getByTestId } = render(
      <ThemeProvider theme={theme}>
        <DndProvider backend={TouchBackend} options={{ enableMouseEvents: true }}>
          {handler.renderEditorContent({
            surveyId: '',
            question: notEmptyQuestion as any,
            onChange,
            confirmOptionRemoval,
          })}
        </DndProvider>
      </ThemeProvider>
    );

    expect(baseElement).toMatchSnapshot();

    const addRowButton = getByTestId('date-time-options-add-row');
    await act(async () => userEvent.click(addRowButton));
    await waitFor(() => {
      expect(onChange).toHaveBeenCalled();
    });
  });

  it('[NEGATIVE] renderEditorContent should return null if question is undefined', async () => {
    const onChange = jest.fn();
    const confirmOptionRemoval = jest.fn();

    expect(
      handler.renderEditorContent({
        surveyId: '',
        question: undefined as any,
        onChange,
        confirmOptionRemoval,
      })
    ).toBeNull();
  });

  it('[NEGATIVE] renderEditorContent should return null if question config is undefined', async () => {
    const onChange = jest.fn();
    const confirmOptionRemoval = jest.fn();

    expect(
      handler.renderEditorContent({
        surveyId: '',
        question: { ...notEmptyQuestion, config: undefined } as any,
        onChange,
        confirmOptionRemoval,
      })
    ).toBeNull();
  });

  it('[NEGATIVE] renderEditorContent should return null if question options are undefined', async () => {
    const onChange = jest.fn();
    const confirmOptionRemoval = jest.fn();

    expect(
      handler.renderEditorContent({
        surveyId: '',
        question: { ...notEmptyQuestion, options: undefined } as any,
        onChange,
        confirmOptionRemoval,
      })
    ).toBeNull();
  });
});
