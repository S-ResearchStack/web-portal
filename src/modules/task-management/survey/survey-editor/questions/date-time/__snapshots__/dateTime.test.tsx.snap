// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`date-time renderEditorContent should render editor content 1`] = `
.c0 {
  height: 9.375rem;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  row-gap: 1rem;
}

.c6 {
  fill: #4475E3;
  cursor: pointer;
  opacity: 0;
  pointer-events: none;
}

.c5 {
  height: 3.5rem;
  width: 12.5rem;
  padding: 1rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  background-color: #F7F8FA;
  border-radius: 0.25rem;
  border: 0.0625rem solid #F7F8FA;
  font-family: Inter,sans-serif;
  font-size: 1rem;
  line-height: 1.5rem;
  font-weight: 400;
  font-style: normal;
}

.c5 svg {
  fill: #808080;
}

.c2 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c7 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-column-gap: 1rem;
  column-gap: 1rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-column-gap: 1.5rem;
  column-gap: 1.5rem;
}

.c8 {
  margin-top: 2rem;
}

.c9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  cursor: pointer;
}

.c10 {
  fill: #4475E3;
  margin-right: 0.625rem;
}

.c11 {
  font-family: Inter,sans-serif;
  font-size: 1rem;
  line-height: 1.5rem;
  font-weight: 600;
  font-style: normal;
  color: #4475E3;
}

<body>
  <div>
    <div
      class="c0"
    >
      <div
        class="c1"
      >
        <div
          class="c2"
        >
          <div
            class="c3"
          >
            <div
              class="c4"
            >
              <div
                class="c5"
              >
                Date
                <svg />
              </div>
            </div>
            <svg
              class="c6"
            />
          </div>
          <div
            class="c7"
          />
        </div>
      </div>
      <div
        class="c8"
      >
        <div
          class="c2"
        >
          <div
            class="c9"
            data-testid="date-time-options-add-row"
          >
            <svg
              class="c10"
            />
            <div
              class="c11"
            >
              Add 
              Time
            </div>
          </div>
          <div
            class="c7"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;
